### ISSUE TITLE
Restore from checkpoint loads optimizer incorrectly

### ISSUE THREAD

**Issue by RomanSteinberg** [tensorflow/tensorflow#52346]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 20.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: -
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): v2.6.0-rc2-32-g919f693420e 2.6.0
- Python version: 3.8
- Bazel version (if compiling from source): -
- GCC/Compiler version (if compiling from source): -
- CUDA/cuDNN version: 11.4
- GPU model and memory: 1080TI 11Gb

**Describe the current behavior**

1. After checkpoint restoration optimizer weights are different from optimizer weights before saving checkpoint.
2. As `assert_consumed` notifies checkpoint file has unresolved optimizer slots (variables). 
```
Unresolved object in checkpoint (root).optimizer.iter: attributes {
  name: "VARIABLE_VALUE"
  full_name: "Adam/iter"
  checkpoint_key: "optimizer/iter/.ATTRIBUTES/VARIABLE_VALUE"
}
```

**Describe the expected behavior**

1. The optimizer weights should be the same before save and after load.
2. The `assert_consumed` should not warn about anything.

**[Contributing](https://www.tensorflow.org/community/contribute)**

- Do you want to contribute a PR? (yes/no): no
- Briefly describe your candidate solution(if contributing): - 

**Standalone code to reproduce the issue**
The reproducible code example is presented in [colab](https://colab.research.google.com/drive/1R01obneq7_jSfBRdUabTAMQIYxuzShqe?usp=sharing).

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.

**Comment by RomanSteinberg** [comment]
Sorry! @tilakrayal Here [it](https://colab.research.google.com/drive/1R01obneq7_jSfBRdUabTAMQIYxuzShqe?usp=sharing) is and I edited colab link in previous post.

**Comment by RomanSteinberg** [comment]
@tilakrayal I read that issue thread and found two workarounds but not the solution. Workarounds are
**1. Use `predict` method to warm up the model before loading weights.** I don't think that it is ok to predict some garbage before loading weights. I'm sure Tensorflow is not designed to work this way.  
**2. Use `tf.Variable` to specify Adam parameters when construct optimizer.** There is no information that Adam optimizer should be constructed like. Moreover default construct behavior is to specify `float`, but not `tf.Variable`. I'm sure Tensorflow is not designed to work this way.

So, please, tell me if I missed something in that issue. Please, tell me if my code contradicts with some TF design, TF pattern or some usage guide. If not, please, admit that it is a bug. I'm ready to contribute to TF to fix it. But I need instructions on what way it should be fixed not to break any idea in one of TF modules.

**Comment by wangpengmit** [comment]
Looping in @k-w-w, who may know more about the checkpointing issue.

**Comment by ghylander** [comment]
I'm having this issue too
My workflow is:
0. (import libs, declare functions, hyperparameters, datasets, checkpoint paths, etc)
1. build model
2. load checkpoints
3. train
4. build model unfreezing layers
5. load best checkpoints of previous training stage
6. train
I get the error right after starting stage 6

**Comment by pcish** [comment]
This is delayed restoration (https://www.tensorflow.org/guide/checkpoint#delayed_restorations) in action. If an additional `train_step()` is called after `checkpoint.restore()`, `assert_consumed()` will report no errors. You can also observe, for example, that `optimizer.iterations` is 11 after the additional `train_step()` -- the value of 10 was restored and then incremented by 1 for the additional `train_step()`.

Hope that helps explain what is going on! While delayed restoration is a necessary feature for eager execution, the group is aware that it can be surprising and unintuitive (even for us...), and there are discussions on what can be done about it.

**Comment by RomanSteinberg** [comment]
@pcish thank you! But don't you think that `assert_consumed` is designed to verify the correct loading? And if it fails, then developer should understand that something is wrong. But you say no it is ok. I think it is much more then surprising...

**Comment by pcish** [comment]
You ask a great question about what `assert_consumed()` is _designed_ to do. Searching through our archives, this is what the original author wrote:

> This allows fully dynamic dependencies, but leads to some possibly surprising corner cases: deferred restoration on object A, creation and modification of variables in object B, then a new dependency from A to B will overwrite the modifications to variables in B with values restored from the checkpoint.
To help users avoid cases like this and for easier debugging, restore will return a status object which can assert loading is complete and no further assignments will take place (assert_consumed())

Elsewhere, the original author provides this example usage code:
```
status = root.restore(tf.train.latest_checkpoint(checkpoint_directory))
for _ in range(num_training_steps):
  ...
status.assert_consumed()
root.save(checkpoint_prefix)
```

Digesting this information, I feel the original author's intention isn't fully coming through. What do you think?

**Comment by RomanSteinberg** [comment]
@pcish I would like to highlight you this [guide](https://www.tensorflow.org/api_docs/python/tf/train/Checkpoint). It has code and suggestions you mentioned, but also suggests to use it after restore. Consider [this part](https://i.imgur.com/L5RlcKF.png).

> This is delayed restoration (https://www.tensorflow.org/guide/checkpoint#delayed_restorations) in action. 
1. Could you suggest a good way to force restoration and avoid problems I described? 
2. Don't you think that `assert_consumed` should force restoration because the developer wishes to verify the restored weights?

**Comment by pcish** [comment]
> @pcish I would like to highlight you this [guide](https://www.tensorflow.org/api_docs/python/tf/train/Checkpoint). It has code and suggestions you mentioned, but also suggests to use it after restore. Consider [this part](https://i.imgur.com/L5RlcKF.png).

Thanks for the links, I'll work on updating the documentation to try to be more clear, and probably not recommend assert_consumed().

> 
> > This is delayed restoration (https://www.tensorflow.org/guide/checkpoint#delayed_restorations) in action.
> 
> 1. Could you suggest a good way to force restoration and avoid problems I described?

You could try running a "fake" training step (e.g. set loss to 0) to force a complete restoration.

> 2. Don't you think that `assert_consumed` should force restoration because the developer wishes to verify the restored weights?

I have to disagree. I'd be rather upset if I wrote:
```
x = has_something_happened()  // false
assert_true(x)
```
and `assert_true(x)` went ahead and `does_something()` because clearly that's what I wished for.

However, I think what you might be getting at is that a `try_consume_everything_now()` function would be helpful?

I'd also like to take a step back and ask if you can share what your high-level goals are that led you to use `assert_consumed()`? Maybe there are other ways to achieve the goal that does not involve a forced restoration or `assert_consumed()`?

**Comment by RomanSteinberg** [comment]
@pcish thank you! These are quite clear answers. I agree both points (1-2).

My high-level goals are typical. Train model (experiments), pick out trained model and deploy it into production service. Unfortunately, I use TF irregularly and cann't follow all trends it suggests from time to time. I mean TF changes usage best practices from time to time. A year or two ago I tried to use TFX, but it was almost impossible to use at that moment. The problem which led me to this issue was to take Checkpoint and make tow things: code for inference in production service and use it later in training. 

a) I tried to load Checkpoint and convert it into SavedModel for inference. It is quite tricky, so I stopped. Decided to use Checkpoint in production service. So, I need to load it and start inference, but it led to use ridiculous code like `model.predict(np.zeros(shape))` after load. I tried to use `model.build`, but as I can see it was removed from guides and is not recommended so. 

b) On other side I'm trying to completely load checkpoint before continue training, and there are some problems with loading optimizer weights (as I showed in this issue). I've noticed incompleteness and started to use `assert_consumed` and ...

PS: A bit of my thoughts. The TF usage became not user friendly. It was easy in v1.0-1.14 and idea to take keras in is the last good step towards simplification of usage.

**Comment by pcish** [comment]
Thanks for sharing your use case and the issues you encountered @RomanSteinberg!

Regarding the original question, I found there's already an open internal feature request for adding an option to disable delayed restoration while loading a checkpoint, which is roughly equivalent to adding a `try_consume_everything_now()` function, so I've added a reference to this conversation on that feature request.

A follow up on your use case: it sounds like you have the source code for a model, and also a Checkpoint for it. Also from the example you gave at the start, I assume you are using Keras. Loading a Checkpoint with the model's source code, then calling `tf.keras.models.save_model()` with the model should export the source code and the weights from the Checkpoint as a SavedModel (that can be used for inference). Was this what you tried but found something tricky?
```
net = load(ckpt_dir, False)  # change load to return net instead of weights
tf.keras.models.save_model(net, '/tmp/saved_model')

inference_model = tf.keras.models.load_model('/tmp/saved_model')
```

**Comment by RomanSteinberg** [comment]
@pcish Yes, I tried to. But SavedModel format is not saving the whole training state, for example optimizer state. So, it is not suitable in my case.

**Comment by pcish** [comment]
I've logged the feature request and updated related documentation, so going to close this issue. Please reopen if you think there is something still missing from the documentation.

**Comment by gengzhanguo** [comment]
> @sanatmpa1 , I was able to reproduce the issue in tf [v2.6](https://colab.research.google.com/gist/tilakrayal/dde9d622b807c51fb5aaa77c76d36c00/2-6checkpoint_save_load_test-1.ipynb), and in [v2.5](https://colab.research.google.com/gist/tilakrayal/1e04a6162b7df86530933b8fb91c8801/2-5checkpoint_save_load_test-1.ipynb) and [nightly](https://colab.research.google.com/gist/tilakrayal/aee3994394cb13aaf5322ac93d0b7515/nightly-checkpoint_save_load_test-1.ipynb) it is failing with different error.Please find the gist here.

Has this problem been solved? I tried this in tensorflow 2.16 and still exists, maybe by saving as HDF5 is better?
